add 6d indexing [WIP]

JaredCrean2 opened the pull request:
This is my first stab at adding a performance test for JuliaLang/julia#18774, using a 6 dimensional array.  I have a script that auto generates these functions, so I can create as high dimension cases as you want.  I haven't created a benchmark before, so let me know if I need to make any changes.

vchuravy commented:
If you can automatically create these functions, maybe it might be worthwhile to put it into a macro?

jrevels commented on src/array/ArrayBenchmarks.jl:
```diff
@@ -235,4 +235,19 @@ g["Float64", "Int"] = @benchmarkable  perf_convert!($x_float, $x_int)
+include("indexing.jl")
+
+g = addgroup!(SUITE, "High Dimensional Indexing", ["6d"])
+dims = (10, 10, 10, 10, 10, 2)  # trailing dimension must be 2
```
Can we change this to something smaller, say `2` for every dimension? Discussion explaining why can be found [here](https://github.com/JuliaCI/BaseBenchmarks.jl/pull/32#issuecomment-254615441), but I should probably also put this advice in the README for future contributors.

jrevels commented on src/array/ArrayBenchmarks.jl:
```diff
@@ -235,4 +235,19 @@ g["Float64", "Int"] = @benchmarkable  perf_convert!($x_float, $x_int)
+
+g = addgroup!(SUITE, "High Dimensional Indexing", ["6d"])
+dims = (10, 10, 10, 10, 10, 2)  # trailing dimension must be 2
+u_i = rand(dims...)
```
`rand` --> `samerand` so that the values here are deterministic (shouldn't really matter for this, but it's one less source of variance to worry about in general).

jrevels commented on src/array/ArrayBenchmarks.jl:
```diff
@@ -235,4 +235,19 @@ g["Float64", "Int"] = @benchmarkable  perf_convert!($x_float, $x_int)
+##########################################
+include("indexing.jl")
+
+g = addgroup!(SUITE, "High Dimensional Indexing", ["6d"])
```
Instead of making a new group for this, can you just add it to [the pre-existing `"index"` group](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/master/src/array/ArrayBenchmarks.jl#L48)?

JaredCrean2 commented:
I added a way to auto-generate `hdindexing.jl`, so now we can test arbitrary dimension arrays.  The generator script was set up to write to file rather than as a macro, so the code generates the file and then includes it.  

I believe I addressed all the comments.  I specified the number of points in each dimension for the 2 through 7 dimension arrays such that a run takes approximately 1/1000 of a second.  The minimum allowable dimension is 5, because it uses a 5 point finite difference stencil.

I left the maximum dimension at 7, but I think indexing should remain efficient for arrays up to MAX_TUPLETYPE_LEN - 2?

jrevels commented on src/array/ArrayBenchmarks.jl:
```diff
@@ -69,6 +69,33 @@ for b in values(g)
+nmax = 6  # maximum dimensionality is nmax + 1
+fname = "hdindexing.jl"
+make_stencil(fname, nmax)  # generate source file
+include("hdindexing.jl")
```
This file no longer exists.

JaredCrean2 replied:
It is generated by `make_stencil`

jrevels replied:
Ah. Do you know where the Travis failure is coming from then? Perhaps the `include` executes before `make_stencil`?

JaredCrean2 replied:
I'll look into it.  I suspect what is happening is the file gets written to the pwd of the shell that launched julia, rather than the directory `ArrayBenchmarks.jl` is in.

jrevels commented on src/array/generate_kernel.jl:
```diff
@@ -0,0 +1,286 @@
+
+
+function getStencil(maxdim::Integer, stencil, dim::Integer, src_eq::Integer, 
+                    second_line_indent::ASCIIString="")
```
Can you change `ASCIIString` to `String`?

JaredCrean2 commented:
That got it.  Good to merge?

jrevels commented on src/array/ArrayBenchmarks.jl:
```diff
@@ -69,6 +69,37 @@ for b in values(g)
+file_rev = reverse(@__FILE__)
+idx = findfirst(file_rev, '/')
+dirpath = reverse(file_rev[idx:end])
+fname = dirpath*"hdindexing.jl"
```
Pretty nifty way of doing this, but might want to just use `dirname` + `joinpath`?

```julia
fname = joinpath(dirname(@__FILE__), "hdindexing.jl")
```

JaredCrean2 replied:
I was wondering if there was an easier way...

jrevels commented:
Thanks again!
